aya/src/programs/info.rs

```diff
@@ -0,0 +1,548 @@
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn id(&self) -> Option<NonZeroU32> {
+        NonZeroU32::new(self.0.id)
+    }
+
+    /// The program tag.
+    ///
+    /// The tag is a SHA sum of the program's instructions which be used as an alternative to
+    /// [`Self::id()`]. A program's ID can vary every time it's loaded or unloaded, but the tag
+    /// will remain the same.
+    ///
+    /// `None` is returned if the field is not available.
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn tag(&self) -> Option<NonZeroU64> {
+        NonZeroU64::new(u64::from_be_bytes(self.0.tag))
+    }
+
+    /// The size in bytes of the program's JIT-compiled machine code.
+    ///
+    /// Note that this field is only updated when BPF JIT compiler is enabled. Kernels v4.15 and
+    /// above may already have it enabled by default.
+    ///
+    /// `None` is returned if the field is not available, or if the JIT compiler is not enabled.
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn size_jitted(&self) -> Option<NonZeroU32> {
+        NonZeroU32::new(self.0.jited_prog_len)
+    }
+
+    /// The size in bytes of the program's translated eBPF bytecode.
+    ///
+    /// The translated bytecode is after it has been passed though the verifier where it was
+    /// possibly modified by the kernel.
+    ///
+    /// `None` is returned if the field is not available.
+    ///
+    /// Introduced in kernel v4.15.
+    pub fn size_translated(&self) -> Option<NonZeroU32> {
+        NonZeroU32::new(self.0.xlated_prog_len)
+    }
+
+    /// The time when the program was loaded.
+    ///
+    /// `None` is returned if the field is not available.
+    ///
+    /// Introduced in kernel v4.15.
+    pub fn loaded_at(&self) -> Option<SystemTime> {
+        if self.0.load_time > 0 {
+            Some(boot_time() + Duration::from_nanos(self.0.load_time))
+        } else {
+            None
+        }
+    }
+
+    /// The user ID of the process who loaded the program.
+    ///
+    /// `None` is returned if the field is not available.
+    ///
+    /// Introduced in kernel v4.15.
+    pub fn created_by_uid(&self) -> Option<u32> {
+        // This field was introduced in the same commit as `load_time`.
+        if self.0.load_time > 0 {
+            Some(self.0.created_by_uid)
+        } else {
+            None
+        }
+    }
+
+    /// The IDs of the maps used by the program.
+    ///
+    /// `None` is returned if the field is not available.
+    ///
+    /// Introduced in kernel v4.15.
+    pub fn map_ids(&self) -> Result<Option<Vec<NonZeroU32>>, ProgramError> {
+        if FEATURES.prog_info_map_ids() {
+            let mut map_ids = vec![0u32; self.0.nr_map_ids as usize];
+            bpf_prog_get_info_by_fd(self.fd()?.as_fd(), &mut map_ids)?;
+
+            Ok(Some(
+                map_ids
+                    .into_iter()
+                    .map(|id| NonZeroU32::new(id).unwrap())
```

Comment:
this should be .expect with a string that mentions the version detection above.

Reply:
Now that I think about it, I think I might of turned the API into an unwrapping nightmare with all the `Option<NonZero*>`. 😵

I'm currently thinking of having fields that were introduced in the same kernel version as the struct not be options (just the direct value). What do you think? 

For `map_ids()`, it looks like it's reasonable to assume the map id would never be zero based on what I traced:
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/syscall.c#L1590: how map IDs are getting filled in `bpf_prog_get_info_by_fd()`
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/verifier.c#L4735: how map info is filled from `bpf_verifier_env` into `bpf_prog_aux`
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/verifier.c#L4249: how map is set in `bpf_verifier_env` before it's verified

Some of the code is shifted in kernel v6.10, but the logic should still be the same. https://elixir.bootlin.com/linux/v6.10/source/kernel/bpf/verifier.c#L18565

cc'ing @alessandrod for thoughts as well

Reply:
> I'm currently thinking of having fields that were introduced in the same kernel version as the struct not be options (just the direct value). What do you think?

Yep this sounds good.

> For map_ids(), it looks like it's reasonable to assume the map id would never be zero based on what I traced:

I don't think that any bpf ids can be 0 regardless of type (map, program, etc)?

In fact I'd probably change all these to return Option<u32>. I think NonZero* is nice to accept as input to enforce safety, but as a return value in these utils seems more annoying than anything else.

Reply:
> I don't think that any bpf ids can be 0 regardless of type (map, program, etc)?

It should not be possible yes

> In fact I'd probably change all these to return Option. I think NonZero* is nice to accept as input to enforce safety, but as a return value in these utils seems more annoying than anything else.

Yeah I do agree with you, I would prefer `Option` over `NonZero*` on returns too.